### PR TITLE
Don't crash when logo passed as integer

### PR DESCRIPTION
**Pull request by Daniel15**
`http://localhost:8080/appveyor/ci/gruntjs/grunt.svg?logo=1` crashes the process at the moment:

```
1203222712 Server is starting up: http://[::]:8080/
1203222728 Vendor hook error: TypeError: s.slice is not a function
    at hasPrefix (C:\src\shields\lib\badge-data.js:20:31)
    at prependPrefix (C:\src\shields\lib\badge-data.js:26:14)
    at makeLogo (C:\src\shields\lib\badge-data.js:59:24)
    at makeBadgeData (C:\src\shields\lib\badge-data.js:89:11)
    at Object.handler (C:\src\shields\server.js:693:19)
    at Domain.vendorDomain.run (C:\src\shields\lib\request-handler.js:170:22)
    at Domain.run (domain.js:242:14)
    at C:\src\shields\lib\request-handler.js:169:18
    at gotQueries (C:\src\shields\node_modules\camp\lib\camp.js:909:7)
    at getQueries (C:\src\shields\node_modules\camp\lib\camp.js:240:5)
C:\src\shields\lib\badge-data.js:20
  return s !== undefined && s.slice(0, prefix.length) === prefix;
                              ^

TypeError: s.slice is not a function
    at hasPrefix (C:\src\shields\lib\badge-data.js:20:31)
    at prependPrefix (C:\src\shields\lib\badge-data.js:26:14)
    at makeLogo (C:\src\shields\lib\badge-data.js:59:24)
    at makeBadgeData (C:\src\shields\lib\badge-data.js:89:11)
    at Timeout.setTimeout [as _onTimeout] (C:\src\shields\lib\request-handler.js:129:25)
    at ontimeout (timers.js:471:11)
    at tryOnTimeout (timers.js:306:5)
    at Timer.listOnTimeout (timers.js:266:5)
```

The failure is contained when running inside the vendorDomain, but then the "vendor servers are unresponsive" code kicks in and doesn't have any error handling, so it crashes the process.

### CODE REVIEW

**Comment by RedSparr0w** [comment]
Nice catch,

Would it be better to catch it inside the `hasPrefix()` function though?
possibly just convert to string with something like:
```javascript
function hasPrefix(s, prefix) {
  return s !== undefined && ('' + s).slice(0, prefix.length) === prefix;
}
```

**Comment by paulmelnikow** [comment]
Thanks @Daniel15. I believe this is the same bug @platan mentioned to me over email last week. Correction: The one he mentioned is different, though also a crasher in `badge-data.js`.

This looks like a bug in my "named logos" code. It seems strange to me that the query parameter ends up being passed in as a number!

This should have a test or two.

@RedSparr0w Thankfully we have [`String#startsWith`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith) now, so I'd rather we removed `hasPrefix`.

**Comment by Daniel15** [comment]
@paulmelnikow  - I think this is somewhere where static typing (Flow or TypeScript) would help. If everything was statically typed, we'd know that this variable might not be a string, so it would complain about the `slice` call (either that or it'd complain about the call site not passing a `string` as expected)

**Comment by paulmelnikow** [comment]
That makes sense… though if this were caught at runtime, wouldn't we still have a runtime crash?

This seems a pretty surprising behavior. I wonder if there's anything else we could get besides a string. An array? A nested object?

**Comment by Daniel15** [comment]
> though if this were caught at runtime, wouldn't we still have a runtime crash?

Yeah we should have a try-catch too. Any exception thrown in the "vendor servers are unresponsive" block of code just kills the process :(

Flow typing would hopefully avoid this happening at runtime, as it'd statically analyze all the variables and throw an error if any code passes a non-string in (which would be picked up via a linter).

> I wonder if there's anything else we could get besides a string. An array? A nested object?

Yeah I wasn't sure, hence the explicit check for whether it's a string.

**Comment by paulmelnikow** [comment]
> Flow typing would hopefully avoid this happening at runtime, as it'd statically analyze all the variables and throw an error if any code passes a non-string in (which would be picked up via a linter).

Could it statically analyze external framework code?

**Comment by Daniel15** [comment]
> Could it statically analyze external framework code?

Yeah, via one of two ways:

1. External framework code can ship Flow types with it. Flow types [can be written as comments](https://flow.org/en/docs/types/comments/) so that the Flow types can remain in 'plain' JS code that's pushed to npm. This works well if the library author is willing to commit to using Flow.

2. For frameworks that don't have any typing, a [library definition file](https://flow.org/en/docs/libdefs/) can be created to specify the type. This requires no changes to the library itself, as the definition file is separate to the library. There's a repository called [flow-typed](https://github.com/flowtype/flow-typed) that has interface definitions for a number of popular projects.

**Comment by paulmelnikow** [comment]
Gotcha. Thanks for explaining that!

**Comment by grajagandev** [comment]
Would you be interested in software that automatically finds TypeErrors like this? I am building [Fuzz Stati0n](https://fuzzstati0n.com/?id=1319) to do that (free for OSS) - please take a look and consider signing up for our newsletter to keep in touch.

**Comment by Daniel15** [comment]
@grajagandev Does it have any advantages over a static type system like Flow or TypeScript?

**Comment by RedSparr0w** [comment]
We recently removed the `hasPrefix` function in #1376 

Shouldn't have any conflicts with this PR though.

**Comment by paulmelnikow** [comment]
I hadn't seen that. Nice! Let's just get the test added then.

**Comment by grajagandev** [comment]
@Daniel15 We have a dynamic analysis tool for forcing uncaught exceptions during runtime (including TypeErrors but also ReferenceErrors, unhandled rejections, etc) - it works on any Node app. We repeatedly send the app unexpected input, crafted to trigger these kinds of critical errors. So a very different solution from TypeScript or Flow.

**Comment by Daniel15** [comment]
If someone wants to add a test for this, feel free to do so. I'll likely not have any time to work on this again until the new year. 😄

File: lib/badge-data.js [hunk 1]
```diff
@@ -20,7 +20,11 @@ function isDataUri(s) {
 function prependPrefix(s, prefix) {
```
Comment (paulmelnikow) [inline lib/badge-data.js]: `s` is expected to be a string, so if we don't get one, it seems we should either coerce to a string or throw an exception.

File: lib/badge-data.js [hunk 2]
```diff
@@ -59,8 +63,10 @@ function makeLabel(defaultLabel, overrides) {
 }
 
 function makeLogo(defaultNamedLogo, overrides) {
-  const maybeDataUri = prependPrefix(overrides.logo, 'data:');
-  const maybeNamedLogo = overrides.logo === undefined ? defaultNamedLogo : overrides.logo;
+  const logo = 'logo' in overrides ? '' + overrides.logo : undefined;
```
Comment (paulmelnikow) [inline lib/badge-data.js]: There could be a related issue when we index into the known logos, so seems safest to coerce here too.

**Pull request merged**